apps/ex_nvr_web/test/ex_nvr_web/controllers/api/device_controller_test.exs

```diff
@@ -83,20 +167,24 @@ defmodule ExNVRWeb.API.DeviceControllerTest do
       %{device: device_fixture(%{settings: %{storage_address: ctx.tmp_dir}})}
     end
 
-    test "get device", %{conn: conn, device: device} do
+    test "get device (:admin)", %{conn: conn, device: device, admin: admin} do
       response =
         conn
         |> get(~p"/api/devices/#{device.id}")
         |> json_response(200)
 
-      assert device.id == response["id"]
+      assert format_device_info_per_role(admin, device) == response
+    end
 
-      assert device.stream_config.stream_uri == response["stream_config"]["stream_uri"]
-      assert device.stream_config.sub_stream_uri == response["stream_config"]["sub_stream_uri"]
-      assert device.stream_config.filename == response["stream_config"]["filename"]
+    test "get device (:user)", %{conn: conn, device: device} do
+      user = user_fixture(%{role: :user})
+      user_conn = log_in_user_with_access_token(conn, user)
+      response =
+        user_conn
+        |> get(~p"/api/devices/#{device.id}")
+        |> json_response(200)
 
-      assert device.credentials.username == response["credentials"]["username"]
-      assert device.credentials.password == response["credentials"]["password"]
+      assert format_device_info_per_role(user, device) == response
```

Comment:
Let's simplify the tests of limited fields for users with`user` role. Check it only when getting one device and rename the test to `get device with filtered fields`. You need to check that the `credentails`, `settings` and `stream_config` are not returned or equals `nil`.

And fix the merge conflict (see Walid updates)